### ISSUE TITLE
atom improvements

### ISSUE THREAD

**Issue by mpenet** [pixie-lang/pixie#424]
Just for the record, a few things that are "missing":

- [ ] watches
- [ ] validation
- [ ] compare-and-set! 
- [ ] metadata support

**Comment by mpenet** [comment]
watches and validation are part of an inherited abstract class in clojure/java, it's shared by other ref types, we might want to do the same (or not). But I don't know what's the long term goal here, it depends if there is a plan to add other ref types (agents, refs) or not (as they are not so popular in clojure, at least in public projects).

**Comment by thomasmulvaney** [comment]
It might be possible to get away with doing this in pixie itself rather than in rpython. In fact atoms are used in only 2 or 3 places from what I can tell in the underlying rpython code (basically where one might use ^:dynamic) in clojure. Maybe we could redefine it in pixie using deftype and give it all the other bells and whistles quite easily with protocols like IWatch

**Comment by mpenet** [comment]
Yes, it's a possibility, my (very uneducated) guess is that atoms are done with rpython for performance reasons, seems like they would be more lightweight this way as well, at least until jit kicks in. And since they are used in some important code such as stateful transducers for instance, it makes sense. But I might be wrong, I don't know well enough about pixie internals tbh. 
Probably something for @halgari to look into.

**Comment by mpenet** [comment]
btw metadata support is in https://github.com/pixie-lang/pixie/pull/425

**Comment by egregius313** [comment]
Is this still of interest? I would be happy to experiment with adding the other features.

**Comment by thomasmulvaney** [comment]
Your welcome to add anything that tickles your interests. These atom improvements would certainly be great!
